File: apps/web/src/lib/recognition.ts
/**
 * Uploading a display photo and reading what was recognised.
 *
 * **The recognition half is not reachable yet.** `POST /api/v1/images` stores
 * the image and returns `{image_id}` and nothing else. AI-008's recogniser is
 * complete — `app/ai/recognition.py`, with a verdict, per-field confidence and
 * an off-topic rejection path — but it is wired to no route, so there is
 * nothing to call. The task's `{recognizedCode, brand, model, confidence}` does
 * not exist over the wire today.
 *
 * This is written against the shape AI-008 actually produces rather than the
 * task's paraphrase of it, so wiring a route is the only thing left to do.
 * `recognise` posts to the endpoint the route will occupy and reports
 * `unavailable` when it is not there, which is what happens now — a state the
 * UI shows honestly rather than a spinner that never resolves.
 */

/** Mirrors `DisplayVerdict` in `app/models/schemas/recognition.py`. */
export type DisplayVerdict = 'fault_display' | 'not_a_fault_display' | 'unreadable';

/**
 * One field the model claims to have read.
 *
 * `confidence` is meaningless without `value`; the backend rejects that
 * combination outright, and `trusted` mirrors its `trusted_at`.
 */
export interface RecognisedField {
  value: string | null;
  confidence: number;
}

/** Mirrors `FaultRecognitionResult`. */
export interface FaultRecognitionResult {
  verdict: DisplayVerdict;
  fault_code: RecognisedField;
  brand: RecognisedField;
  model: RecognisedField;
  note: string | null;
}

/**
 * The confidence floor, matching `MIN_FIELD_CONFIDENCE` in AI-008.
 *
 * Duplicated rather than imported because it lives in Python. It is asserted
 * against the backend constant in the tests, so the two cannot drift silently.
 */
export const MIN_FIELD_CONFIDENCE = 0.8;

/** What the upload step produced. */
export type UploadOutcome =
  | { kind: 'recognised'; imageId: string; result: FaultRecognitionResult }
  | { kind: 'stored'; imageId: string }
  | { kind: 'failed'; reason: UploadFailure };

export type UploadFailure =
  'too-large' | 'rejected' | 'network' | 'unavailable' | 'timeout' | 'aborted';

/**
 * A field may be used without asking the engineer to confirm it.
 *
 * The verdict is checked first, deliberately. A model that decided the photo
 * is a wiring diagram and reported a code anyway has invented the code, and a
 * caller reading fields before the verdict would use it — which is why the
 * backend schema refuses that combination outright.
 */
export function trusted(
  result: FaultRecognitionResult,
  field: keyof FaultRecognitionResult,
): boolean {
  if (result.verdict !== 'fault_display') return false;
  const value = result[field];
  if (typeof value !== 'object' || value === null) return false;
  const recognised: RecognisedField = value;
  return recognised.value !== null && recognised.confidence >= MIN_FIELD_CONFIDENCE;
}

/**
 * Should the engineer be asked to confirm before a diagnosis is run?
 *
 * High confidence on a readable display pre-fills the message for a one-tap
 * send; anything else surfaces the confirmation prompt instead. The asymmetry
 * is deliberate: a wrong fault code sends the engineer to the wrong procedure,
 * and one extra tap is a much smaller cost than that.
 */
export function needsConfirmation(result: FaultRecognitionResult): boolean {
  return !trusted(result, 'fault_code');
}

export interface UploadOptions {
  file: File;
  token: string;
  signal?: AbortSignal;
  fetchImpl?: typeof fetch;
  endpoint?: string;
  /**
   * How long before the UI stops claiming progress.
   *
   * The spec asks for explicit slow-upload messaging rather than a silent
   * spinner: on a factory-floor connection a large upload can legitimately
   * take a while, and an engineer cannot tell that from a hung request.
   */
  slowAfterMs?: number;
  onSlow?: () => void;
}

/** Upload one image and, when a recogniser exists, read what it saw. */
export async function uploadImage(options: UploadOptions): Promise<UploadOutcome> {
  const {
    file,
    token,
    signal,
    fetchImpl = fetch,
    endpoint = '/api/v1/images',
    slowAfterMs = 6000,
    onSlow,
  } = options;

  const body = new FormData();
  body.append('file', file);

  const slowTimer = onSlow ? setTimeout(onSlow, slowAfterMs) : undefined;

  try {
    const response = await fetchImpl(endpoint, {
      method: 'POST',
      headers: { Authorization: `Bearer ${token}` },
      body,
      ...(signal ? { signal } : {}),
    });

    if (response.status === 413) return { kind: 'failed', reason: 'too-large' };
    if (response.status === 404) return { kind: 'failed', reason: 'unavailable' };
    if (!response.ok) return { kind: 'failed', reason: 'rejected' };

    const payload: unknown = await response.json();
    const imageId = readImageId(payload);
    if (imageId === null) return { kind: 'failed', reason: 'rejected' };

    // The recogniser's output, if the endpoint has grown one. Absent today,
    // and absence is reported as `stored` rather than as a failure — the
    // image did upload, and the engineer can still describe the fault.
    const result = readRecognition(payload);
    return result ? { kind: 'recognised', imageId, result } : { kind: 'stored', imageId };
  } catch {
    return { kind: 'failed', reason: signal?.aborted ? 'aborted' : 'network' };
  } finally {
    if (slowTimer !== undefined) clearTimeout(slowTimer);
  }
}

/** Read the image id, tolerating a payload that is not the shape we expect. */
function readImageId(payload: unknown): string | null {
  if (typeof payload !== 'object' || payload === null) return null;
  const id = (payload as { image_id?: unknown }).image_id;
  return typeof id === 'string' && id !== '' ? id : null;
}

/**
 * Read a recognition result, if one is present and well formed.
 *
 * Validated rather than cast. The whole point of the verdict is that fields
 * are not read until it has been checked, and a payload that arrived without
 * one would defeat that by being trusted anyway.
 */
function readRecognition(payload: unknown): FaultRecognitionResult | null {
  if (typeof payload !== 'object' || payload === null) return null;
  const raw = (payload as { recognition?: unknown }).recognition;
  if (typeof raw !== 'object' || raw === null) return null;

  const verdict = (raw as { verdict?: unknown }).verdict;
  if (
    verdict !== 'fault_display' &&
    verdict !== 'not_a_fault_display' &&
    verdict !== 'unreadable'
  ) {
    return null;
  }

  const note = (raw as { note?: unknown }).note;
  return {
    verdict,
    fault_code: readField(raw, 'fault_code'),
    brand: readField(raw, 'brand'),
    model: readField(raw, 'model'),
    note: typeof note === 'string' ? note : null,
  };
}

function readField(raw: object, name: string): RecognisedField {
  const field = (raw as Record<string, unknown>)[name];
  if (typeof field !== 'object' || field === null) return { value: null, confidence: 0 };
  const value = (field as { value?: unknown }).value;
  const confidence = (field as { confidence?: unknown }).confidence;
  return {
    value: typeof value === 'string' && value !== '' ? value : null,
    // A confidence that is not a number is treated as none. A field carrying
    // a number it cannot justify is exactly what the backend validator
    // rejects, and defaulting high here would reintroduce it.
    confidence: typeof confidence === 'number' && Number.isFinite(confidence) ? confidence : 0,
  };
}

File: apps/api/app/models/schemas/recognition.py
"""What a vision model may report about a photo of an equipment display.

**Three fields, not a description.** Narrow structured extraction is both more
reliable than "describe this image" and gateable: a confidence attached to
each field lets the fault code be trusted while the brand is not, which is the
common case when a nameplate is out of frame.

**"Not a fault display" is a first-class answer.** A stressed engineer
photographs the wrong thing — a nameplate, a wiring diagram, their own boot.
A model asked only to extract a fault code will find something code-shaped in
almost any image, and a fabricated code sends someone to the wrong procedure
with full confidence. So the model reports whether it is looking at a fault
display at all, and that verdict is checked before any field is read.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field, StringConstraints, model_validator

# A recognised value is a short token off a screen, not prose.
ReadValue = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=100)]


class DisplayVerdict(StrEnum):
    """Whether the photo shows what the engineer thinks it shows."""

    # A fault, alarm or status display with a readable code.
    FAULT_DISPLAY = "fault_display"
    # Equipment, but not a fault display — a nameplate, a terminal strip, a
    # wiring diagram. Recognisable, just not what was asked for.
    NOT_A_FAULT_DISPLAY = "not_a_fault_display"
    # A display that is there but cannot be read: glare, motion blur, angle,
    # a screen that is off. Distinct from the above because the engineer's
    # fix is different — retake the photo, rather than photograph something
    # else.
    UNREADABLE = "unreadable"


class RecognisedField(BaseModel):
    """One extracted value and how sure the model is of it.

    Attributes:
        value: What was read, verbatim off the screen. ``None`` when the field
            is not visible — which is ordinary, not an error: a fault display
            usually shows a code and rarely shows the manufacturer.
        confidence: How sure the model is, in [0, 1]. Meaningless without
            ``value``, and validated as absent in that case.
    """

    value: ReadValue | None = None
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)

    @model_validator(mode="after")
    def _confidence_needs_a_value(self) -> RecognisedField:
        """Keep a confidence from standing on its own.

        Returns:
            The validated field.

        Raises:
            ValueError: If confidence is claimed for a field that was not
                read. A high confidence next to an absent value is the shape
                of an answer, and something downstream will eventually read
                the number without checking the value beside it.
        """
        if self.value is None and self.confidence != 0.0:
            raise ValueError("confidence without a value means nothing")
        return self

    def trusted_at(self, threshold: float) -> bool:
        """Report whether this field may be used without confirmation.

        Args:
            threshold: The confidence floor.

        Returns:
            ``True`` if a value was read and clears the floor. A field with no
            value is never trusted, whatever the number beside it.
        """
        return self.value is not None and self.confidence >= threshold


class FaultRecognitionResult(BaseModel):
    """What the model saw.

    Attributes:
        verdict: Whether this is a readable fault display at all. Checked
            before any field is read.
        fault_code: The code shown, e.g. "F0001".
        brand: The manufacturer, if visibly identifiable — usually only when a
            logo is in frame.
        model: The equipment model, if visibly identifiable.
        note: What the model saw when the verdict is not ``FAULT_DISPLAY``,
            so the engineer is told what to do instead of just "no".
    """

    verdict: DisplayVerdict
    fault_code: RecognisedField = Field(default_factory=RecognisedField)
    brand: RecognisedField = Field(default_factory=RecognisedField)
    model: RecognisedField = Field(default_factory=RecognisedField)
    note: str | None = None

    @model_validator(mode="after")
    def _a_rejected_photo_reads_nothing(self) -> FaultRecognitionResult:
        """Keep a rejection from carrying extracted values.

        Returns:
            The validated result.

        Raises:
            ValueError: If a non-fault-display verdict carries a fault code.
                That combination is the exact failure this schema exists to
                prevent: a model that has decided the photo is a wiring
                diagram and reported a code anyway has invented the code, and
                a caller reading fields before the verdict would use it.
        """
        if self.verdict is not DisplayVerdict.FAULT_DISPLAY and self.fault_code.value is not None:
            raise ValueError(
                f"verdict is {self.verdict.value} but a fault code was reported; "
                "a code read off something that is not a fault display is invented"
            )
        return self
